README.md
# fastify-opencensus

[![Node version](https://img.shields.io/node/v/fastify-opencensus)]()
[![Downloads Count](https://img.shields.io/npm/dm/fastify-opencensus)]()
[![Build Status](https://travis-ci.org/rhaymo/fastify-opencensus.svg?branch=master)](https://travis-ci.org/rhaymo/fastify-opencensus)
[![Known Vulnerabilities](https://snyk.io//test/github/rhaymo/fastify-opencensus/badge.svg?targetFile=package.json)](https://snyk.io//test/github/rhaymo/fastify-opencensus?targetFile=package.json)
[![Coverage Status](https://coveralls.io/repos/github/rhaymo/fastify-opencensus/badge.svg?branch=master)](https://coveralls.io/github/rhaymo/fastify-opencensus?branch=master)
[![License](https://img.shields.io/github/license/rhaymo/fastify-opencensus)](https://github.com/rhaymo/fastify-opencensus/blob/master/LICENSE)

Metrics collector for Fastify based on [Opencensus](https://opencensus.io/) .

This module is inspired and based on the [fastify-metrics](https://gitlab.com/m03geek/fastify-metrics) plugin.

This plugin adds 3 http metrics for your routes:

- Requests duration distribution
- Requests duration summary
- Requests count

## ToC

- [fastify-opencensus](#fastify-opencensus)
  - [ToC](#ToC)
  - [Fastify support](#Fastify-support)
  - [Installation](#Installation)
  - [Features and requirements](#Features-and-requirements)
  - [Usage](#Usage)
    - [Plugin options](#Plugin-options)
      - [Metrics details](#Metrics-details)
    - [HTTP routes metrics](#HTTP-routes-metrics)
  - [Docs](#Docs)
  - [Changelog](#Changelog)
  - [See also](#See-also)
  - [License](#License)

## Fastify support

- **v0.x.x** - will support `>= fastify-2.0.0`

## Installation

```sh
npm i fastify-opencensus --save
```

<sub>[Back to top](#toc)</sub>

## Features and requirements

- Collects default server metrics (see [opencensus-default-metrics](https://github.com/rhaymo/opencensus-node-default-metrics));
- Collects route response timings
- By default, metrics info are collected into the global instance of opencensus required by this package. If you want to use another opencensus version, you have to pass in this instance using the `stats` options field.

---

- Requires fastify `>=2.0.0`.
- Node.js `>=8.9.0`.

<sub>[Back to top](#toc)</sub>

## Usage

Add it to your project like regular fastify plugin. Use `register` method and pass options to it.

```js
const fastify = require('fastify');
const app = fastify();

const metricsPlugin = require('fastify-opencensus');
app.register(metricsPlugin, { interval: 3000 });
```

It also exports opencensus stats instance to fastify instance `fastify.opencensus.client` which you may use it in your routes.

See example folder for other examples.

<sub>[Back to top](#toc)</sub>

### Plugin options

| parameter              | type                     | description                                                                     | default      |
| ---------------------- | ------------------------ | ------------------------------------------------------------------------------- | ------------ |
| `enableDefaultMetrics` | Boolean                  | Enables collection of node default metrics.                                     | `true`       |
| `enableStats`          | Boolean                  | Enables collection of fastify metrics.                                          | `true`       |
| `stats`                | Object                   | Custom opencensus metrics instance.                                             | `undefined`  |
| `metricsExporter`      | Array<Object>            | Array of Opencensus metrics exporter                                            | `undefined`  |
| `groupStatusCodes`     | Boolean                  | Groups status codes (e.g. 2XX) if `true`                                        | `false`      |
| `pluginName`           | String                   | Change name which you'll use to access opencensus registry instance in fastify. | `opencensus` |
| `interval`             | Number                   | Default metrics collection interval in ms.                                      | `5000`       |
| `blacklist`            | String, RegExp, String[] | Skip metrics collection for blacklisted routes                                  | `undefined`  |
| `prefix`               | String                   | Custom default metrics prefix.                                                  | `""`         |
| `metrics`              | Object                   | Allows override default metrics config. See section below.                      | `{}`         |

#### Metrics details

You may override default metrics settings. You may provide overrides for three metrics tracking http request durations, count and sum (labelNames cannot be overriden).
Default values:

```js
{
  distribution: {
    name: 'http_request_duration_seconds',
    desc: 'request duration in seconds',
    labelNames: ['status_code', 'method', 'route'],
    buckets: [0.05, 0.1, 0.5, 1, 3, 5, 10],
  },
  sum: {
    name: 'http_sum_request_duration_seconds',
    desc: 'Sum of durations of http requests',
    labelNames: ['status_code', 'method', 'route']
  },
  count: {
    name: 'http_request_count',
    desc: 'Counter of http requests',
    labelNames: ['status_code', 'method', 'route']
  },
}
```

Override should look like:

```js
const fastify = require('fastify');
const app = fastify();
const metricsPlugin = require('fastify-metrics');

app.register(metricsPlugin, {metrics: {
  distribution: {
    name: 'my_custom_http_request_duration_seconds',
    buckets: [0.1, 0.5, 1, 3, 5],
  },
  sum: {
    desc: 'custom request duration in seconds summary help'
  },
});
```

<sub>[Back to top](#toc)</sub>

### HTTP routes metrics

The following table shows what metrics will be available in Prometheus. Note suffixes like `_bucket`, `_sum`, `_count` are added automatically.

| metric                              | labels                           | description                  |
| ----------------------------------- | -------------------------------- | ---------------------------- |
| `http_request_count`                | `method`, `route`, `status_code` | Requests total count         |
| `http_request_duration_seconds`     | `method`, `route`, `status_code` | Requests durations by bucket |
| `http_sum_request_duration_seconds` | `method`, `route`, `status_code` | Requests duration sum        |

<sub>[Back to top](#toc)</sub>

## Docs

See [docs](docs/README.md).

<sub>[Back to top](#toc)</sub>

## Changelog

See [changelog](CHANGELOG.md).

<sub>[Back to top](#toc)</sub>

## See also

- [opencensus-default-metrics](https://github.com/rhaymo/opencensus-node-default-metrics) - collector of node default metrics using opencensus

<sub>[Back to top](#toc)</sub>

## License

Licensed under [MIT](./LICENSE).

<sub>[Back to top](#toc)</sub>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.2](https://github.com/rhaymo/fastify-opencensus/compare/v0.0.1...v0.0.2) (2019-09-25)

### 0.0.1 (2019-09-20)


example/test.js
const path = require('path');
const metricsPlugin = require('../dist/index');
const fastifyBuilder = require('fastify');

function init() {
  const fastify = fastifyBuilder({
    // logger: { level: conf.LogLevel }
  });
    /*
    fastify.register(fastifyShutdown).after((err) => {
        fastify.log.error(err);
        // Register custom clean up handler
        fastify.gracefulShutdown((code, cb) => {
            cb();
        });
    }); */

  const {PrometheusStatsExporter} = require('@opencensus/exporter-prometheus');
  const exporter = new PrometheusStatsExporter({
    port: 9464,
    startServer: true,
  });
  fastify.register(metricsPlugin, {endpoint: '/metrics', metricsExporter: [exporter]});

  fastify.ready(() => {
    const routes = fastify.printRoutes();
    // log.debug(logCatalog.FASTIFY_AVAILABLE_ROUTES, routes);
  });

  fastify.get('/info', {
    schema: {hide: true},
  }, async (request, reply) => {
    reply.send('ok');
  });


  fastify.listen(process.env.PORT || 3000, '0.0.0.0');
  console.log('started');
}


process.on('unhandledRejection', (reason) => {
  throw reason;
});
process.on('uncaughtException', (error) => {
  console.log(error);
});

init();


test/stat.spec.ts
import 'jest';

import OpenCensusMetrics from '../src/stat';
import { globalStats } from '@opencensus/core';

describe('OpenCensusMetrics', () => {
    test('createMetrics', async () => {
        const openCensusMetrics: OpenCensusMetrics = new OpenCensusMetrics(globalStats);
        openCensusMetrics.createMetrics({}, 'myprefix_' );
        const metrics = globalStats.getMetrics();
      expect(metrics).toHaveLength(3);

      expect(metrics[0].descriptor.description).toEqual('The distribution of the latencies');
      expect(metrics[0].descriptor.name).toEqual('myprefix_http_request_duration_seconds');
    });
});


test/util.spec.ts
import 'jest';

import { collectMetricsForUrl , sinceInMilliseconds} from '../src/util';

describe('util', () => {
    describe('collectMetricsForUrl', () => {

        test('empty blacklists should return true', async () => {
            const url='/blacklisted';
            expect(collectMetricsForUrl(undefined, url)).toEqual(true);
        });

        test('blacklisted string should return false', async () => {
            const blacklist: string = '/blacklisted';
            const url='/blacklisted?bla=po';
            expect(collectMetricsForUrl(blacklist, url)).toEqual(false);
        });

        test('blacklisted url in Array should return false', async () => {
            const blacklist: string = '/blacklisted';
            const url='/blacklisted';
            expect(collectMetricsForUrl([blacklist], url)).toEqual(false);
        });

        test('blacklisted url matching a regexp should return false', async () => {
            const url='/blacklisted';
            expect(collectMetricsForUrl(new RegExp('black'), url)).toEqual(false);
        });

        test('not blacklisted string should return true', async () => {
            const blacklist: string = 'blacklisted';
            const url='/notblacklisted';
            expect(collectMetricsForUrl(blacklist, url)).toEqual(true);
        });

        test('blacklisted url in Array should return true', async () => {
            const blacklist: string = '/blacklisted';
            const url='/notblacklisted';
            expect(collectMetricsForUrl([blacklist], url)).toEqual(true);
        });

        test('blacklisted url matching a regexp should return true', async () => {
            const url='/whitelisted';
            expect(collectMetricsForUrl(new RegExp('black'), url)).toEqual(true);
        });

    });

    describe('sinceInMilliseconds', () => {
        test('blacklisted url matching a regexp should return true', async () => {
            const url='/whitelisted';
            expect(typeof sinceInMilliseconds(process.hrtime()[1])).toBe('number');
        });
    });
});


src/index.ts
import * as http from 'http';
import { FastifyInstance, Plugin } from 'fastify';
import fastifyPlugin from 'fastify-plugin';
import { globalStats} from '@opencensus/core';
import { PluginOptions, FastifyOpenCensus } from './plugin';
import { collectMetricsForUrl, sinceInMilliseconds } from './util';
import OpenCensusMetrics from './stat';
import {collectDefaultMetrics} from 'opencensus-default-metrics';

declare module 'fastify' {
  interface FastifyInstance<
    HttpServer = http.Server,
    HttpRequest = http.IncomingMessage,
    HttpResponse = http.ServerResponse
    > {
    /**
     * Metrics interface
     */
    metrics: FastifyOpenCensus;
  }
  interface RouteSchema {
    /**
     * Hides metric route from swagger/openapi documentation
     */
    hide?: boolean; // for compatibility with fastify-oas
  }
  interface FastifyRequest<HttpRequest, Query, Params, Headers, Body> {
    metrics?: {
      /**
       * Request start time
       */
      startTime: number;

    };
  }
}

/**
 * Fastify OpenCensus plugin
 */
const fastifyOpenCensusPlugin: Plugin<
  http.Server,
  http.IncomingMessage,
  http.ServerResponse,
  PluginOptions
> = function fastifyOpenCensus(
  fastify: FastifyInstance,
  {
    enableDefaultMetrics = true,
    enableStats = true,
    stats = globalStats,
    metricsExporter,
    groupStatusCodes = false,
    pluginName = 'opencensus',
    interval = 5000,
    blacklist,
    prefix,
    metrics = {}
  }: PluginOptions = {},
  next: fastifyPlugin.nextCallback
) {
    const openCensusMetrics: OpenCensusMetrics = new OpenCensusMetrics(stats, metricsExporter);

    if (enableStats) {
      openCensusMetrics.createMetrics(metrics, prefix || '');

      fastify.addHook('onRequest', (request, _, next) => {
        if (request.req.url && collectMetricsForUrl(blacklist, request.req.url)) {
          request.metrics = {
            startTime: process.hrtime()[1]
          };
        }
        next();
      });

      fastify.addHook('onResponse', function (request, reply, next) {
        if (request.metrics) {
          let routeId = reply.context.config.url || request.req.url;
          if (reply.context.config.statsId) {
            routeId = reply.context.config.statsId;
          }
          const method = request.req.method;
          const statusCode = groupStatusCodes
            ? `${Math.floor(reply.res.statusCode / 100)}xx`
            : reply.res.statusCode;

          openCensusMetrics.recordLatencyMeasurement(sinceInMilliseconds(request.metrics.startTime), method || 'UNKNOWN', statusCode, routeId);
        }
        next();
      });
    }

    if (enableDefaultMetrics) {
      const defaultOpts = {
        timeout: interval,
        stats : stats,
        prefix
      };
      collectDefaultMetrics(defaultOpts);
    }

    const plugin: FastifyOpenCensus = { client: stats };
    plugin.clearRegister = stats.clear;
    fastify.decorate(pluginName, plugin);

    next();
  };

export = fastifyPlugin(fastifyOpenCensusPlugin, {
  fastify: '>=2.0.0',
  name: 'fastify-opencensus',
});


src/plugin.ts
import { Stats, StatsEventListener } from '@opencensus/core';

export interface HTTPMetric {
  /**
   * Metric name
   */
  name?: string;
  /**
   * Metric description
   */
  desc?: string;
  /**
   * Histogram/Summary buckets
   */
  buckets?: Array<number>;
}

export interface Metrics {
  /**
   * http request durations
   */
  distribution?: HTTPMetric;

  /**
   * http request count
   */
  count?: HTTPMetric;

  /**
   * http request duration sum
   */
  sum?: HTTPMetric;
}

export interface FastifyOpenCensus {
  /**
   * global Stats of opencensus
   */
  client: Stats;

  /**
   * Expose register clear function if register was provided
   */
  clearRegister?(): void;
  /**
   * Additional objects to store your metrics, registries, etc.
   */
  [key: string]: any;
}

export interface PluginOptions {
  /**
   * Enable default nodejs metrics
   * @default true
   */
  enableDefaultMetrics?: boolean;

  /**
   * Enable stats metrics for http route
   * @default true
   */
  enableStats?: boolean;

  /**
   * Groups status code labels by first digit 200 -> 2XX
   * @default false
   */
  groupStatusCodes?: boolean;
  /**
   * Plugin name that will be registered in fastify
   * @default opencensus
   */
  pluginName?: string;
  /**
   * Metrics collection interval in ms
   * @default 5000
   */
  interval?: number;
  /**
   * Routes blacklist that will be excluded from metrics collection
   */
  blacklist?: RegExp | Array<string> | string;
  /**
   * opencensus registry for default metrics and route metrics
   */
  stats?: Stats;

  /**
   * Metrics prefix
   */
  prefix?: string;
  /**
   * Metrics endpoint for Prometheus
   */
  endpoint?: string;
  /**
   * HTTP metrics overrides
   */
  metrics?: Metrics;

  metricsExporter?: Array<StatsEventListener>;
}


example/example2_tracing_zPage.js
// How to set multiple metric exporters and a zPage traces exporter

const metricsPlugin = require('../dist/index');
const fastify = require('fastify')();

// The block 7-16 MUST precedes every other require of @opencensus. Try to move line 18 before line 7 and the tracing won't work
const tracing = require('@opencensus/nodejs');
tracing.start();

const zpages = require('@opencensus/exporter-zpages');
const options = {
  port: 8083,
  startServer: true,
};
const exporterZpages = new zpages.ZpagesExporter(options);
tracing.registerExporter(exporterZpages);

const {PrometheusStatsExporter} = require('@opencensus/exporter-prometheus');
const exporter = new PrometheusStatsExporter({
  port: 9464,
  startServer: true,
});
const exporter1 = new PrometheusStatsExporter({
  port: 9465,
  startServer: true,
  prefix: 'anotherProm',
});
// eslint-disable-next-line max-len
fastify.register(metricsPlugin, {interval: 2000, metricsExporter: [exporter, exporter1]});

fastify.get('/info', {
  schema: {hide: true},
}, (request, reply) => {
  reply.send('ok');
});

process.on('unhandledRejection', (reason) => {
  throw reason;
});
process.on('uncaughtException', (error) => {
  console.log(error);
});

fastify.listen(process.env.PORT || 3000, '0.0.0.0');
console.log('started');



example/example3_tracing_zipkin.js
// How to set multiple metric exporters and a zipkin traces exporter

const metricsPlugin = require('../dist/index');
const fastify = require('fastify')();


const tracing = require('@opencensus/nodejs');
tracing.start();
const {ZipkinTraceExporter} = require('@opencensus/exporter-zipkin');
const zipkinOptions = {
  url: 'http://localhost:9411/api/v2/spans',
  serviceName: 'myZipkinExample2',
};
const zipkinExporter = new ZipkinTraceExporter(zipkinOptions);
tracing.registerExporter(zipkinExporter);

const {PrometheusStatsExporter} = require('@opencensus/exporter-prometheus');
const exporter = new PrometheusStatsExporter({
  port: 9464,
  startServer: true,
});
fastify.register(metricsPlugin, {interval: 2000, metricsExporter: [exporter]});

fastify.get('/info', {
  schema: {hide: true},
}, (request, reply) => {
  reply.send('ok');
});


process.on('unhandledRejection', (reason) => {
  throw reason;
});
process.on('uncaughtException', (error) => {
  console.log(error);
});

fastify.listen(process.env.PORT || 3000, '0.0.0.0');
console.log('started');



src/stat.ts
import {
    Stats, MeasureUnit, AggregationType, TagMap, Measure, View, StatsEventListener
} from '@opencensus/core';
import { Metrics } from './plugin';

const methodTagKey = { name: 'method' };
const statusTagKey = { name: 'status_code' };
const errorTagKey = { name: 'route' };

class OpenCensusMetrics {
    mLatency: Measure;
    stats: Stats;

    constructor(_stats: Stats, metricsExporters?: Array<StatsEventListener>) {
        this.stats = _stats;

        if (metricsExporters) {
            metricsExporters.forEach((metricExporter) => { this.stats.registerExporter(metricExporter); });
        }
    }

    createMetrics(metrics: Metrics, _prefix?: string): void {
        // The latency in milliseconds
        this.mLatency = this.stats.createMeasureDouble(
            'http_request_duration_seconds',
            MeasureUnit.SEC,
            'http request duration in seconds'
        );

        const latencyDistribution: View = this.stats.createView(
            metrics.distribution && metrics.distribution.name || `${_prefix}http_request_duration_seconds`,
            this.mLatency,
            AggregationType.DISTRIBUTION,
            [methodTagKey, statusTagKey, errorTagKey],
            metrics.distribution && metrics.distribution.desc || 'The distribution of the latencies',
            metrics.distribution && metrics.distribution.buckets || [0.05, 0.1, 0.5, 1, 3, 5, 10]
        );
        this.stats.registerView(latencyDistribution);

        const httpRequestCount: View = this.stats.createView(
            metrics.count && metrics.count.name || `${_prefix}http_request_count`,
            this.mLatency,
            AggregationType.COUNT,
            [methodTagKey, statusTagKey, errorTagKey],
            metrics.count && metrics.count.desc || 'Counter of http request'
        );
        this.stats.registerView(httpRequestCount);

        const httpRequestDurationSum: View = this.stats.createView(
            metrics.sum && metrics.sum.name || `${_prefix}http_sum_request_duration_seconds`,
            this.mLatency,
            AggregationType.SUM,
            [methodTagKey, statusTagKey, errorTagKey],
            metrics.sum && metrics.sum.desc || 'Sum of durations of http request'
        );
        this.stats.registerView(httpRequestDurationSum);
    }

    recordLatencyMeasurement(_value: number, method: string, statusCode: string | number, routeId: string): void {
        const tags = new TagMap();
        tags.set(methodTagKey, { value: method || 'UNKNOWN' });
        tags.set(statusTagKey, { value: '' + statusCode });
        tags.set(errorTagKey, { value: routeId });

        this.stats.record([{
            measure: this.mLatency,
            value: _value
        }], tags);
    }

    /*
    startTracing(tracingOptions) {
        this.tracing.start(tracingOptions);
    }

    startZPagesServer(options) {
        const defaultOptions = {
            port: 8081,
            startServer: true
        };

        const _options = Object.assign(defaultOptions, options);

        const zpages = require('@opencensus/exporter-zpages');
        const exporter = new zpages.ZpagesExporter(_options);
        this.tracing.registerExporter(exporter);
    }
    */
}

export = OpenCensusMetrics;
